### PR TITLE
bosh rollout behaviour: delete pods after updating partition

### PR DESCRIPTION
**Pull request by phil9909**
Currently, we delete `Pods` which would block the rollout and then update the `StatefulSet` partition. This has two problems:
* If the update failes (e.g. concurrent modification) we've deleted a pod for no reason.
* There is a potential race: If the `StatefulSet controller` recreates the deleted pod before we update the partition this update it will still use the old (potentially broken) version and thereby potentially block the rollout.

### CODE REVIEW

File: pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go [hunk 1]
```diff
@@ -164,6 +163,18 @@ func (r *ReconcileStatefulSetRollout) Reconcile(request reconcile.Request) (reco
 	if dirty {
 		meltdown.SetLastReconcile(&statefulSet.ObjectMeta, time.Now())
 		err = r.update(ctx, &statefulSet, &result)
+		if err != nil {
+			ctxlog.Debug(ctx, "Error updating StatefulSet ", request.NamespacedName, err)
```
[the remark this answers is not in the document]
Comment (phil9909) [reply pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go]: I don't think so. It is pretty normal to happen when the the resource is stale (i.e. has been updated by the statefulset controller in the meantime). We will just re-queue and try again.

File: pkg/kube/controllers/statefulset/statefulset_rollout_reconciler_test.go [hunk 2]
```diff
@@ -316,6 +317,21 @@ var _ = Describe("ReconcileStatefulSetRollout", func() {
 					Expect(client.UpdateCallCount()).To(Equal(0))
 					Expect(client.DeleteCallCount()).To(Equal(0))
 				})
+
+				When("update fails", func() {
+					JustBeforeEach(func() {
+						client.UpdateReturns(errors.New("injected error"))
+					})
+
+					It("no pod is deleted", func() {
```
[the remark this answers is not in the document]
Comment (c0d1ngm0nk3y) [reply pkg/kube/controllers/statefulset/statefulset_rollout_reconciler_test.go]: Done

File: pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go [hunk 3]
```diff
@@ -164,6 +163,18 @@ func (r *ReconcileStatefulSetRollout) Reconcile(request reconcile.Request) (reco
...
+			return reconcile.Result{}, err
+		}
+
+		if oldPartition != *statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition {
+			err = CleanupNonReadyPod(ctx, r.client, &statefulSet, *statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition)
+			if err != nil {
+				ctxlog.Debug(ctx, "Error calling CleanupNonReadyPod ", request.NamespacedName, err)
```
[the remark this answers is not in the document]
Comment (c0d1ngm0nk3y) [reply pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go]: Done

File: pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go [hunk 4]
```diff
@@ -87,6 +87,13 @@ func (r *ReconcileStatefulSetRollout) Reconcile(request reconcile.Request) (reco
 	var status = statefulSet.Annotations[AnnotationCanaryRollout]
 	var newStatus = status
 	dirty := false
+	var oldPartition int32
+	if statefulSet.Spec.UpdateStrategy.RollingUpdate.Partition == nil {
+		oldPartition = -1
+		ctxlog.Debug(ctx, "StatefulSet with unexpected partition = nil found (webhook should prevent this)", request.NamespacedName)
```
[the remark this answers is not in the document]
Comment (c0d1ngm0nk3y) [reply pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go]: Done

**Pull request merged**